woocommerce/woocommerce-gutenberg-products-block
# Update dynamic block styles and markup for default themes

This PR styles the new dynamic grid blocks across the default WordPress themes, as well as maintaining compatibility with Storefront.

Tested on;
- Twenty Nineteen
- Twenty Seventeen
- Twenty Sixteen
- Twenty Fifteen
- Storefront

Keeping 'core' style class names such as `.products` and `.product` was challenging because there were conflicts with existing styles. To help combat this without resorting to `!important` tags I had to:

1) Increase specificity of our styles (which is why you'll see more nesting of rules in this PR)
2) Force the blocks stylesheet to load later than the core stylesheets (https://github.com/woocommerce/woocommerce-gutenberg-products-block/commit/bc87d5e897fa3d4aae9bd0521d1788216321f455)
   - This was done using a core WooCommerce filter. The other advantage to that is that the block styles can be unhooked like other core styles.
   - Due to this change, I had to remove `style` when registering blocks
   - And again, to keep the styles in the editor I had to add an `@import` rule
3) Add a div to the block fragment/render to match our list based structure so styles were correctly applied. I think this is temporary however as we are moving to server side rendering.
4) Add a body class which indicates theme e.g. `theme-storefront` (https://github.com/woocommerce/woocommerce-gutenberg-products-block/commit/a684cfb6f1ac86cd140a9059037b1a096a1f9359)

Markup wise, I had to make some additional adjustments to allow proper styling:

- Moved the sale badge out of the price wrapper to allow existing Storefront styles to function
- Added missing `price` class
- Added a missing escape on the permalink
- Added wp block button classes (https://github.com/woocommerce/woocommerce-gutenberg-products-block/commit/e23e3dc00c15f6c1fdb483472e8d6851f98981a4)

Our markup does not match core exactly. The placement of the ratings for example are different, and (now) only the title and image are within the link which to me makes sense. Given this could require change by a theme developer, I've wrapped the new HTML with a filter so that HTML can be changed (https://github.com/woocommerce/woocommerce-gutenberg-products-block/commit/7519ae5e802043b12aac6c4a39b9af93bda6b53c).


Fixes #568

After merge, next steps are probably to roll out dynamic blocks everywhere, and to implement the server side rendering. We may also need to do tests under non-core themes to see how much they are impacted.

#### Accessibility

No new styles were added in this PR aside from those affecting layout. The colours etc are all inherited.

### Screenshots

![2015](https://user-images.githubusercontent.com/90977/58025282-07ab4680-7b0c-11e9-8eea-ed59cebaf87c.png)
![2016](https://user-images.githubusercontent.com/90977/58025287-0b3ecd80-7b0c-11e9-93ac-53ef34f46076.png)
![2017](https://user-images.githubusercontent.com/90977/58025293-0da12780-7b0c-11e9-965e-1285f375044a.png)
![2019](https://user-images.githubusercontent.com/90977/58025298-10038180-7b0c-11e9-8faa-adc07ce11056.png)
![Storefront](https://user-images.githubusercontent.com/90977/58025302-11cd4500-7b0c-11e9-8fe6-f89c5cfa69f8.png)

### How to test the changes in this Pull Request:

1. Checkout and test the appearance of the new products/sale block under each of the core themes listed above. 
2. Test using 1,2,3,4 columns - font sizes should change as more columns are added as before.
3. Ensure layout is maintained under Storefront
4. Ensure layout is maintained in the editor
